### ISSUE TITLE
Alacritty Does Not Use GPU In Hybrid Graphics Mode

### ISSUE THREAD

**Issue by bonedaddy** [alacritty/alacritty#3587]
I'm not sure if I would really describe this as a bug, however when my laptop is running in hybrid graphics mode, alacritty doesn't seem to acquire use of the GPU, and instead will use the CPU for rendering.

Here is the output from `nvidia-smi`
```
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 440.64       Driver Version: 440.64       CUDA Version: 10.2     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce GTX 1650    Off  | 00000000:01:00.0 Off |                  N/A |
| N/A   37C    P8     7W /  N/A |     31MiB /  3911MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    0      1726      G   /usr/lib/xorg/Xorg                            14MiB |
|    0      2225      G   /usr/lib/xorg/Xorg                            14MiB |
+-----------------------------------------------------------------------------+
```

Notice that under the processes, it's just Xorg, and the alacrity process isn't listed there

### System

OS: Linux
Version: solidity@gray ~/C/R/Temporal> 
Linux/BSD:Pop!_OS (pretty sure DE)

### Logs

Crashes: not applicable
Font/Terminal size:
```
Created log file at "/tmp/Alacritty-4360.log"
[2020-04-13 16:47] [INFO] Welcome to Alacritty
[2020-04-13 16:47] [INFO] Configuration loaded from "/home/solidity/.alacritty.yml"
[2020-04-13 16:47] [DEBUG] Estimated DPR: 1
[2020-04-13 16:47] [DEBUG] Estimated Cell Size: 9 x 17
[2020-04-13 16:47] [DEBUG] Estimated Dimensions: None
[2020-04-13 16:47] [INFO] Device pixel ratio: 1
[2020-04-13 16:47] [INFO] Initializing glyph cache...
[2020-04-13 16:47] [INFO] ... finished initializing glyph cache in 0.004614572s
[2020-04-13 16:47] [INFO] Cell Size: 9 x 17
[2020-04-13 16:47] [INFO] Padding: 0 x 0
[2020-04-13 16:47] [INFO] Width: 800, Height: 600
[2020-04-13 16:47] [INFO] PTY Dimensions: Line(35) x Column(88)
[2020-04-13 16:47] [INFO] Initialisation complete
[2020-04-13 16:47] [DEBUG] Term::resize dimensions unchanged
[2020-04-13 16:47] [INFO] Width: 800, Height: 600
[2020-04-13 16:47] [DEBUG] [unhandled osc_dispatch]: [['7',],['f','i','l','e',':','/','/','g','r','a','y','/','h','o','m','e','/','s','o','l','i','d','i','t','y','/','C','o','d','e','/','R','T','r','a','d','e','L','t','d','/','T','e','m','p','o','r','a','l',],] at line 758
```
Keyboard and bindings: ive attached as file
[Alacritty-4441.log](https://github.com/alacritty/alacritty/files/4472400/Alacritty-4441.log)

**Comment by chrisduerr** [comment]
Alacritty should never use the CPU for rendering, unless the OpenGL implementation is specifically told to do so.

I'd assume you mean Alacritty uses the integrated GPU, instead of the dedicated one?

If you disable the integrated GPU, do you then see it use the dedicated GPU successfully?

Could you try changing the following line to take `Some(true)` instead of `None` to see if that works?

https://github.com/alacritty/alacritty/blob/master/alacritty/src/window.rs#L120

Alacritty by default doesn't require the dedicated GPU, I believe the reasoning behind that was saving power. However that decision was made back in 2018 so I don't quite remember (see #1403).

**Comment by bonedaddy** [comment]
Hey thanks for the response, I think I misspoke by saying "using the cpu", and meant to say the CPU's integrated GPU processor or w.e. the things intel CPUs use to do rendering.

> If you disable the integrated GPU, do you then see it use the dedicated GPU successfully?

Yea I switched to to using nvidia graphics instead of hybrid graphics and `nvidia-smi` shows alacritty process:

```
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 440.64       Driver Version: 440.64       CUDA Version: 10.2     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce GTX 1650    Off  | 00000000:01:00.0 Off |                  N/A |
| N/A   45C    P5    13W /  N/A |    491MiB /  3911MiB |     15%      Default |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    0      1660      G   /usr/lib/xorg/Xorg                            27MiB |
|    0      2212      G   /usr/lib/xorg/Xorg                           203MiB |
|    0      2420      G   /usr/bin/gnome-shell                          78MiB |
|    0      3148      G   alacritty                                     10MiB |
|    0      3953      G   ...AAAAAAAAAAAAAAgAAAAAAAAA --shared-files    63MiB |
|    0      4712      G   ...uest-channel-token=....................    36MiB |
|    0      5950      G   ...quest-channel-token=...................    23MiB |
+-----------------------------------------------------------------------------+
```

> Could you try changing the following line to take Some(true) instead of None to see if that works?
> https://github.com/alacritty/alacritty/blob/master/alacritty/src/window.rs#L120

Sure I can  try that tomorrow

**Comment by stokhos** [comment]
@chrisduerr changing None to Some(true) works.

**Comment by chrisduerr** [comment]
Hmm okay. So on macOS with our application settings having it set to `None` just means that it will use the iGPU on battery and the dGPU when plugged in. I feel like that's probably the ideal approach, but I'm not quite sure how to achieve that on Linux/Windows.

Just to confirm, when testing this you are running on wall power, not battery, right?

Alacritty's primary goal is performance, so it seems tough to justify picking anything other than the optimal performance options by default. However that's useless when you cannot use the application because it immediately drains all battery. So I think that dGPU on wallpower is really the best choice we have.

Alternatively I could imagine adding a configuration option for this. Something that can be set to `true`/`false`/`None` maybe, where `true` will always enable the dGPU and `None` will only activate it when plugged in the wall.

**Comment by stokhos** [comment]
Yes, my laptop were running on wall power.  I totally understand this. I'm ok with the default configs, but it would be nice if there was an option for this in configuration.

**Comment by nixpulvis** [comment]
I too would like to see a configuration option. However I'm not sure exactly how we should expose this. macOS has the "Automatic graphics switching" system preference, which enables the use of the iGPU. I'm not sure what the Windows/X11/Wayland equivalents are at this time, nor how this works across various GPU vendors.

I think I'd expect this behaviour from Alacritty however:

- If there is more than one enabled GPU, use the fastest when connected to power.
- If there is still more than one GPU when disconnected from power, use the best power/performace ratio device. Alacritty is very performant, so a slower GPU should still work just fine, and saving battery by default seems smart.
- If there is only one GPU, well... use it.

I personally own a desktop (always connected to power) and an Apple laptop which have both an integrated and dedicated GPU, so I could help test / investigate this issue a bit.

**Comment by chrisduerr** [comment]
Alacritty doesn't pick the GPU it runs on. So this configuration is really just to have an option to force enable dGPUs, force disable them, or use automatic graphics switching.

**Comment by chrisduerr** [comment]
@bonedaddy Do you think you could test the linked PR? Unfortunately I don't have a machine with switchable graphics.

**Comment by mrkvn** [comment]
In PopOS, I also use Hybrid Graphics. I run alacritty using the following command based on [this](https://support.system76.com/articles/graphics-switch-pop/):
`__NV_PRIME_RENDER_OFFLOAD=1 __GLX_VENDOR_LIBRARY_NAME=nvidia alacritty`

Now my alacritty is running on Nvidia.

I bind the command to a shortcut so that I don't have to open gnome-terminal to run the command. Apparently, can't make the command to bind on keyboard shortcut under Settings. I used [Simple X Hotkey Daemon](https://github.com/baskerville/sxhkd) for binding.

**Comment by chrisduerr** [comment]
@mrkvn Have you tried the linked PR (#4133)?

**Comment by mrkvn** [comment]
@chrisduerr I tried. Not sure if I did it right. I added the following config to my .yml file:
```
# Power management mode
  #
  # This will control things like the GPU used to render Alacritty. When set to
  # `Auto`, Alacritty will automatically switch to low power mode when on
  # battery.
  #
  # Values for `energy`:
  #   - Performance
  #   - Powersave
  #   - Auto
  energy: Performance
```

Didn't work. Let me know if I did it wrong.

**Comment by chrisduerr** [comment]
What laptop setup do you have exactly? Does it support *automatic* graphics switching in any application?

**Comment by mrkvn** [comment]
```
OS: Pop!_OS 20.10 x86_64
Kernel: 5.8.0-7630-generic
GPU: NVIDIA GeForce GTX 1050 Mobile
GPU: Intel HD Graphics 630
```
```
$ sudo system76-power graphics
hybrid
```

I'm not so knowledgeable but I think automatic graphics switching is enabled. I've setup a specific game using Lutris and Nvidia is being used when I open the game, otherwise, it's not in use. I know that the hybrid switching is working because on initial install of the OS, it only uses Nvidia, which outputs a noticeable heat on my laptop and now, it's not.

**Comment by chrisduerr** [comment]
@mrkvn You're aware that the `energy` setting is below the `window` parent, right?

So it should look something like this:

```yml
window:
  energy: Performance
```

**Comment by mrkvn** [comment]
> @mrkvn You're aware that the `energy` setting is below the `window` parent, right?
> 
> So it should look something like this:
> 
> ```yaml
> window:
>   energy: Performance
> ```

Yeah. I'm not aware at the time I posted my comment earlier. But I did try that, the correct one, under `window`. Still not opening using Nvidia.

Sorry. I'm really new to this. Correct me if I'm wrong. What I did is clone your repo, checkout to commit 8f75519, build it. Uncomment the `window` and `energy` portion from the provided yml file. Take note that I didn't remove the other comments, so it looked like:
```
window:

# ...
# some comments here
# ...

  energy: Performance

```
I ran it using the command:
`./<some-path>/alacritty/target/release/alacritty --config-file <some-path>/alacritty/alacritty.yml`

Ran `nvidia-smi` and didn't find alacritty.

**Comment by chrisduerr** [comment]
There should be an easy way to test that the config option is getting read and the correct commit was pulled: Change it from `Performance` to something like `aoeu`, then it should print an error telling you the available options.

**Comment by mrkvn** [comment]
> There should be an easy way to test that the config option is getting read and the correct commit was pulled: Change it from `Performance` to something like `aoeu`, then it should print an error telling you the available options.

Got this:
```
[2021-01-04 05:09:05.917827459] [ERROR] [alacritty_config] Problem with config: unknown variant `blabla`, expected one of `Performance`, `Powersave`, `Auto`; using default value
Created log file at "/tmp/Alacritty-59737.log"
```

Looks like it's being read properly. It could be OS specific problem. It's hard to know. In my setup, at least, it's not working. The OP should test it so that you'd get another feedback.

**Comment by chrisduerr** [comment]
You're running linux, so it should work. The original patch was tested, but never with configuration options. Unfortunately I don't have access to any machines with switchable graphics and linux. But thanks for trying at least.

**Comment by mrkvn** [comment]
Sure. Glad to be of help, which actually didn't help :)

**Comment by jorgebef** [comment]
Hi all,
I've been trying to support this too and I've just now encountered something that's a bit weird.
I'm on a MBP 16" with 5300M graphics, all set to automatic graphics switching.
When I launch Alacritty application it won't switch to dedicated gpu even with `energy: Performance` set accordingly, however, if I launch alacritty from a terminal, it will switch to the dedicated gpu.
I have no idea what's going on, how could I help to try to see what's going on under the hood?
Thank you for this great project!

**Comment by mrkvn** [comment]
I also tested `energy: Performance` in Arch Linux using Optimus Manager, didn't work too.

**Comment by xsrvmy** [comment]
Ran into this as well when trying to figure out if nvidia is faster by any chance, not that it matters too much because it's already way faster than konsole.
I tried using the __GLX... and __NV_PRIME... flags. This will make glxinfo report nvidia, but it still doesn't use the nvidia gpu.

Could this have something to do with

**Comment by uahic** [comment]
On Ubuntu 22 I installed Alacritty and followed the window: energy: Powersafe as suggested here, did not work. Currently my computer is freshly setup and I'm still using the gnome standard desktop thing of Ubuntu, when launching Alacritty by the menubar on the left side I noticed that it offers a dedicated option 'Launch using discrete graphics card' and that brought me to the solution. Ubuntu 22 uses the utility 'switcherooctl' to decide which GPU to use 

In my case:

``` 
➜  ~ switcherooctl 
Device: 0
  Name:        Intel Corporation TigerLake-H GT1 [UHD Graphics]
  Default:     yes
  Environment: DRI_PRIME=pci-0000_00_02_0

Device: 1
  Name:        NVIDIA Corporation GA107GLM [RTX A2000 Mobile]
  Default:     no
  Environment: __GLX_VENDOR_LIBRARY_NAME=nvidia __NV_PRIME_RENDER_OFFLOAD=1

```
One way to use it: 
``` alias dgpu="export __GLX_VENDOR_LIBRARY_NAME=nvidia && export __NV_PRIME_RENDER_OFFLOAD=1 &&" ``` in your bash/zsh/... config and use it for your keybindings to pop up a terminal. If someone has a more clever approach to set it as a default for Alacritty, go a head :)

**Comment by ajukraine** [comment]
I just want to add to what @jorgebef said. When Alacritty is started from within Kitty terminal, it stars using dedicated GPU on my MacBook Pro 16.

There is more details for my investigation here: https://github.com/neovim/neovim/issues/21989

**Comment by kchibisov** [comment]
That's because you need settings coming from the `Alacritty.app` that's how macOS works. The preferences to decide which GPU to pick are in plists.
